eth/watchers/pricefeedwatcher.go

```diff
@@ -0,0 +1,176 @@
+type PriceFeedWatcher struct {
+	ctx            context.Context
+	baseRetryDelay time.Duration
+
+	priceFeed                   eth.PriceFeedEthClient
+	currencyBase, currencyQuote string
+
+	current      eth.PriceData
+	priceUpdated chan eth.PriceData
+}
+
+// NewPriceFeedWatcher creates a new PriceFeedWatcher instance. It will already
+// fetch the current price and start a goroutine to watch for updates.
+func NewPriceFeedWatcher(ctx context.Context, rpcUrl, priceFeedAddr string) (*PriceFeedWatcher, error) {
+	priceFeed, err := eth.NewPriceFeedEthClient(ctx, rpcUrl, priceFeedAddr)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create price feed client: %w", err)
+	}
+
+	description, err := priceFeed.Description()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get description: %w", err)
+	}
+
+	currencyFrom, currencyTo, err := parseCurrencies(description)
+	if err != nil {
+		return nil, err
+	}
+
+	w := &PriceFeedWatcher{
+		ctx:            ctx,
+		baseRetryDelay: priceUpdateBaseRetryDelay,
+		priceFeed:      priceFeed,
+		currencyBase:   currencyFrom,
+		currencyQuote:  currencyTo,
+		priceUpdated:   make(chan eth.PriceData, 1),
+	}
+
+	err = w.updatePrice()
+	if err != nil {
+		return nil, fmt.Errorf("failed to update price: %w", err)
+	}
+
+	go func() {
+		ctx, cancel := context.WithCancel(w.ctx)
+		defer cancel()
+		ticker := newTruncatedTicker(ctx, priceUpdatePeriod)
+		w.watch(ctx, ticker)
+	}()
+
+	return w, nil
+}
+
+// Currencies returns the base and quote currencies of the price feed.
+// i.e. base = CurrentPrice() * quote
+func (w *PriceFeedWatcher) Currencies() (base string, quote string) {
+	return w.currencyBase, w.currencyQuote
+}
+
+// Current returns the latest fetched price data.
+func (w *PriceFeedWatcher) Current() eth.PriceData {
+	return w.current
+}
+
+// PriceUpdated returns a channel that will receive the updated price data as
+// soon as it changes.
+func (w *PriceFeedWatcher) PriceUpdated() <-chan eth.PriceData {
+	return w.priceUpdated
+}
+
+func (w *PriceFeedWatcher) updatePrice() error {
+	newPrice, err := w.priceFeed.FetchPriceData()
+	if err != nil {
+		return fmt.Errorf("failed to fetch price data: %w", err)
+	}
+
+	if newPrice.UpdatedAt.After(w.current.UpdatedAt) {
+		w.current = newPrice
+		select {
+		case w.priceUpdated <- newPrice:
+		default:
+		}
+	}
+
+	return nil
+}
+
+func (w *PriceFeedWatcher) watch(ctx context.Context, ticker <-chan time.Time) {
+	for {
+		select {
+		case <-w.ctx.Done():
+			return
+		case <-ticker:
+			attempt, retryDelay := 1, w.baseRetryDelay
+			for {
+				err := w.updatePrice()
+				if err == nil {
+					break
+				} else if attempt >= priceUpdateMaxRetries {
+					clog.Errorf(ctx, "Failed to fetch updated price from PriceFeed attempts=%d err=%q", attempt, err)
+					break
+				}
+
+				clog.Warningf(ctx, "Failed to fetch updated price from PriceFeed, retrying after retryDelay=%d attempt=%d err=%q", retryDelay, attempt, err)
+				time.Sleep(retryDelay)
```

Comment:
I don't love the sleep here, because it effectively blocks the whole for loop. E.g. if someone closes the context, it will stop only after this sleep. How about changing it to another ticker/timer?

Anyway, if you decide to keep this sleep, I'm ok with this. Shouldn't be too harmful :)

Reply:
Was relatively simple to fix by switching to a `time.After` instead. Done :)